README.md

# Spinal Cord Toolbox [![Build Status](https://travis-ci.org/neuropoly/spinalcordtoolbox.svg?branch=master)](https://travis-ci.org/neuropoly/spinalcordtoolbox)

Comprehensive and open-source library of analysis tools for MRI of the spinal cord.

- [Installation](https://sourceforge.net/p/spinalcordtoolbox/wiki/installation/)
- [List of tools](http://sourceforge.net/p/spinalcordtoolbox/wiki/tools/)
- [Getting started](https://github.com/neuropoly/spinalcordtoolbox/blob/master/batch_processing.sh)
- [List of changes](https://github.com/neuropoly/spinalcordtoolbox/blob/master/CHANGES.md)

## References

#### Spinal Cord Toolbox
- [Cohen-Adad et al. Spinal Cord Toolbox: an open-source framework for processing spinal cord MRI data. Proceedings of the 20th Annual Meeting of OHBM, Hamburg, Germany 2014:3633](https://www.dropbox.com/s/liravffu5b4rdni/cohenadad_ohbm14_spinalcordtoolbox_v2.pdf?dl=0)
- [De Leener et al. Template-based analysis of multi-parametric MRI data with the Spinal Cord Toolbox. Proc. ISMRM, Toronto, Canada 2015](https://www.dropbox.com/s/1gv81xg17lgcxfn/5798_submitted_radC862D.pdf)

#### Segmentation (Propseg)
- [De Leener et al. Robust, accurate and fast automatic segmentation of the spinal cord. Neuroimage 2014;98:528-536.](https://dl.dropboxusercontent.com/u/20592661/publications/DeLeener_NIMG14_PropSeg.pdf)
- [Ullmann et al. Automatic labeling of vertebral levels using a robust template-based approach. Int J Biomed Imaging 2014;Article ID 719520.](http://downloads.hindawi.com/journals/ijbi/2014/719520.pdf)
- [De Leener et al. Automatic segmentation of the spinal cord and spinal canal coupled with vertebral labeling. IEEE Transactions on Medical Imaging 2015; (in press)](https://www.dropbox.com/s/2tsp1juxcvuq4bl/PID3727605.pdf)

#### Template and Atlas
- [Fonov et al. Framework for integrated MRI average of the spinal cord white and gray matter: The MNI-Poly-AMU template. Neuroimage 2014;102P2:817-827.](https://dl.dropboxusercontent.com/u/20592661/publications/Fonov_NIMG14_MNI-Poly-AMU.pdf)
- [Taso et al. Construction of an in vivo human spinal cord atlas based on high-resolution MR images at cervical and thoracic levels: preliminary results. MAGMA, Magn Reson Mater Phy 2014;27(3):257-267](https://dl.dropboxusercontent.com/u/20592661/publications/taso_magma_2014.pdf)
- [Lévy et al. White matter atlas of the human spinal cord with estimation of partial volume effect. Neuroimage 2015 (in press); doi: 10.1016/j.neuroimage.2015.06.040](http://www.sciencedirect.com/science/article/pii/S1053811915005431)
- [Cadotte et al. Characterizing the Location of Spinal and Vertebral Levels in the Human Cervical Spinal Cord. AJNR Am J Neuroradiol 2014;36(5):1-8.](https://dl.dropboxusercontent.com/u/20592661/publications/Cadotte_AJNR14_SpinalLevel.pdf)
- [Touati et al. T1 and T2 template of the brainstem and spinal cord. Proc. ISMRM, Toronto, Canada 2015](https://www.dropbox.com/s/i7at66hugucrs1h/touati_ismrm15_template.pdf)

#### Registration
- [De Leener et al. Template-based analysis of multi-parametric MRI data with the Spinal Cord Toolbox. Proc. ISMRM, Toronto, Canada 2015](https://dl.dropboxusercontent.com/u/20592661/publications/deleener_ismrm15_sct.pdf) 
- [Cohen-Adad et al. Slice-by-slice regularized registration for spinal cord MRI: SliceReg. Proc. ISMRM, Toronto, Canada 2015](https://dl.dropboxusercontent.com/u/20592661/publications/cohenadad_ismrm15_slicereg.pdf)

## License

The MIT License (MIT)

Copyright (c) 2014 École Polytechnique, Université de Montréal

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.




CHANGES.md
# CHANGES TO RELEASE

##(TBD)
- BUG: **install_patch**: now possible to install as non-admin (issues #380, #434)
- 

##2.0.6 (2015-06-30)
- BUG: **sct_process_segmentation**: fixed bug of output file location (issue #395)

##2.0.5 (2015-06-10)
- BUG: **sct_process_segmentation**: fixed error when calculating CSA (issue #388)

##2.0.4 (2015-06-06)
- BUG: **sct_process_segmentation**: fixed error when calculating CSA (issue #388)
- BUG: Hanning smoothing: fixed error that occurred when window size was larger than data (issue #390)
- OPT: **sct_check_dependences**: now checks if git is installed
- OPT: simplified batch_processing.sh

##2.0.3 (2015-05-19)
- BUG: **sct_register_to_template**: fixed issue related to appearance of two overlapped templates in some cases (issue #367)
- BUG: **sct_register_to_template**: now all input data are resampled to 1mm iso to avoid label mismatch (issue #368)
- BUG: **sct_resample**: fixed bug when user specified output file
- OPT: **sct_create_mask**: improved speed

##2.0.2 (2015-05-16)
- BUG: **sct_fmri_compute_tsnr**: fixed issue when input path includes folder
- BUG: **sct_orientation**: now possibility to change orientation even if no qform in header (issue #360)
- BUG: **msct_smooth**: fixed error with small Hanning window (issue #363)
- BUG: **sct_straighten_spinalcord**: fixed issue with relative path (issue #365)
- NEW: **sct_label_utils**: added new function to transform group of labels into discrete label points
- NEW: **sct_orientation**: added a tool to fix wrong orientation of an image (issue #366)
- OPT: **sct_register_to_template**: twice as fast! (issue #343)

##2.0.1 (2015-04-28)
- BUG: **sct_extract_metric**: MAP method did not scale properly with the data. Now fixed (issue #348)
- BUG: fixed issue with parser when typing a command to see usage (it crashed)

##2.0 (2015-04-17)

- NEW: **sct_fmri_compute_tsnr**: new function to compute TSNR from fMRI data (performs moco before)
- OPT: **sct_straighten_spinalcord**: now MUCH faster and more accurate (issue #240)
- OPT: **sct_register_to_template**: allows more flexibility by allowing multiple steps for registration (flag -p).
  - N.B. flag "-m" has been replaced by "-s"
- OPT: **sct_register_multimodal**: allows more flexibility by imposing only one stage. Several stages can be run sequentially and then transformations can be concatenated.
  - N.B. flags "-s" and "-t" were replaced with "-iseg" and "-dseg" respectively
- OPT: **sct_extract_metric**: 
  - new methods for extraction: maximum likelihood and maximum a posteriori, which take into account partial volume effect
  - now possible to specify global regions for extraction with flag -l: wm, gm, sc
  - now possible to include a bunch of labels using ":". Example: 2:29
- NEW: **sct_get_centerline_from_labels**: obtain a centerline using a combination of labels and/or segmentations
  - N.B. sct_get_centerline was renamed for sct_get_centerline_automatic
- NEW: **sct_compute_ernst_angle**: new script to compute and display Ernst angle depending on T1 and TR
- OPT: **sct_process_segmentation**:
  - can compute average CSA across vertebral levels or slices
  - can compute length of segmentation
  - can compute CSA on non-binary images such as probabilistic gray/white matter maps
  - N.B. process names were simplified to: "csa", "length" and "centerline"
- OPT: **sct_crop_image**: now possible to crop an image based on a reference space
- OPT: new WM atlas: added gray matter and CSF for computing partial volume
- OPT: now use all available cores for ANTs and adjust variable when running dmri_moco (issue #238)
- INST: new installer in python, simpler to use and check for latest patches
- REF: msct_parser: new parser that generate documentation/usage
- REF: msct_image, sct_label_utils: smoothly converting the toolbox to objet-oriented, some scripts can be used as python module

##1.1.2_beta (2014-12-25)

- BUG: sct_dmri_moco: fixed crash when using mask (issue # 245)
- OPT: sct_create_mask: (1) updated usage (size in vox instead of mm), (2) fixed minor issues related to mask size.
- INST: links are now created during installation of release or patch (issue ).

##1.1.1 (2014-11-13)

- FIX: updated ANTs binaries for compatibility with GLIBC_2.13 (issue: https://sourceforge.net/p/spinalcordtoolbox/discussion/help/thread/e00b2aeb/)

##1.1 (2014-11-04)

- NEW: sct_crop: function to quickly crop an image.
- NEW: sct_extract_metric (replaces the old sct_estimate_MAP_tracts.py). New functionalities added (maximum likelihood estimation and tract grouping). More flexible with label files.
- NEW: sct_convert_mnc2nii
- NEW: sct_create_mask: create mask of different shapes (cylinder, box, gaussian). Useful for moco.
- NEW: sct_fmri_moco: motion correction function for fMRI data. Uses regularization along z.
- NEW: sct_compute_mtr: compute MTR
- NEW: sct_otsu: OTSU segmentation (usefull for DWI data)
- NEW: sct_resample: quick upsample/downsample 3D or 4D data
- NEW: sct_segment_greymatter: function to segment the grey matter by warping that one from the atlas
- OPT: sct_orientation can now be applied to 4d data
- OPT: sct_register_multimodal now using the new antsSliceReg method that regularizes along z.
- OPT: new version of the white matter atlas: more accurate, deformation accounting for internal structure (use BSplineSyN instead of SyN).
- OPT: sct_dmri_moco now using the new antsSliceReg method that regularizes along z.
- OPT: removed all .py extensions for callable functions (created links)
- OPT: sct_label_utils: now possible to create labels. Also added other useful features.
- INST: now possible to specify installation path for the toolbox
- INST: conda dependences are now automatically installed by the installer.
- INST: added pillow (fixed issue #117)
- INST: "getting started" now provided via example commands in batch_processing.sh
- REF: sct_straighten_spinalcord (fixed issues #56, #116)

sct_class/CreateMask.py
#!/usr/bin/env python
#########################################################################################
#
# sct_create_mask obect equivalent. Creates a mask along the Z axis
#
# ---------------------------------------------------------------------------------------
# Copyright (c) 2014 Polytechnique Montreal <www.neuro.polymtl.ca>
# Authors: Benjamin Leener, Julien Cohen-Adad
# Modified: 2015-05-13
#
# About the license: see the file LICENSE.TXT
#########################################################################################

import sys
from msct_parser import Parser
import os
import getopt
import commands
from msct_image import Image
import sct_utils as sct
import time
import numpy
import nibabel
from scipy import ndimage
from sct_orientation import get_orientation, set_orientation
from Algorithm import Algorithm


class Mask(Algorithm):

    # class constants definition
    OUTPUT_PREFIX = "mask_"
    METHOD_LIST = ['coord', 'point', 'centerline', 'center']
    SHAPE_LIST = ['cylinder', 'box', 'gaussian']
    DEFAULT_SHAPE = "cylinder"
    DEFAULT_METHOD = "center"
    DEFAULT_SIZE = 41

    def __init__(self, input_image, output_file=None, method=DEFAULT_METHOD, method_value=None, shape=DEFAULT_SHAPE, size=DEFAULT_SIZE, verbose=1, rm_tmp_files=1, rm_output_file=0, produce_output=1):
        super(Mask, self).__init__(input_image, produce_output, verbose)
        self._output_file = output_file
        self._method = method
        self._method_value = method_value
        self._shape = shape
        self._size = size
        self._rm_tmp_files = rm_tmp_files
        self._result = None
        self._rm_output_file = rm_output_file

    @property
    def output_file(self):
        return self._output_file

    @output_file.setter
    def output_file(self, value):
        self._output_file = value

    @property
    def method(self):
        return self._method

    @method.setter
    def method(self, value):
        self._method = value

    @property
    def method_value(self):
        return self._method_value

    @method_value.setter
    def method_value(self, value):
        self._method_value = value

    @property
    def shape(self):
        return self._shape

    @shape.setter
    def shape(self, value):
        self._shape = value

    @property
    def size(self):
        return self._size

    @size.setter
    def size(self, value):
        self._size = value

    @property
    def rm_tmp_files(self):
        return self._rm_tmp_files

    @rm_tmp_files.setter
    def rm_tmp_files(self, value):
        self._rm_tmp_files = value

    @property
    def result(self):
        return self._result

    @result.setter
    def result(self, value):
        self._result = value

    @property
    def rm_tmp_files(self):
        return self._rm_tmp_files

    @rm_tmp_files.setter
    def rm_tmp_files(self, value):
        self._rm_tmp_files = value

    def mask_method(self, nx, ny, nz, nt, px, py, pz, pt):
        raise NotImplementedError("Use a child class of mask to chose which method will be used")

    def execute(self):
        """
        This method creates a mask and returns the result either through a file or as an msct_image Image object

        """
        fsloutput = 'export FSLOUTPUTTYPE=NIFTI; '

        method_type = self._method

        # TODO : Erase when used with parser
        if not method_type == 'center':
            method_val = self._method_value
            if method_type == 'centerline':
                sct.check_file_exist(method_val, self.verbose)

        # checking if shape is valid
        if self._shape not in Mask.SHAPE_LIST:
            sct.printv('\nERROR in '+os.path.basename(__file__)+': Shape "'+self._shape+'" is not recognized. \n', self.verbose, 'error')

        # checking orientation in RPI
        if not get_orientation(self.input_image) == 'RPI':
            sct.printv('\nERROR in '+os.path.basename(__file__)+': Orientation of input image should be RPI. Use sct_orientation to put your image in RPI.\n', 1, 'error')

        # display input parameters
        sct.printv('\nInput parameters:', self.verbose)
        sct.printv('  data ..................'+self.input_image, self.verbose)
#        sct.printv('  method ................'+method_type, self.verbose)

        # Extract path/file/extension
        path_data, file_data, ext_data = sct.extract_fname(self.input_image)

        # Get output folder and file name
        if self._output_file is None:
            self._output_file = Mask.OUTPUT_PREFIX+file_data+ext_data

        # creating temporary folder
        path_tmp = sct.slash_at_the_end('tmp.'+time.strftime("%y%m%d%H%M%S"), 1)
        os.mkdir(path_tmp)

        # Copying input data to tmp folder and convert to nii
        # NB: cannot use c3d here because c3d cannot convert 4D data.
        sct.printv('\nCopying input data to tmp folder and convert to nii...', self.verbose)
        sct.run('cp '+self.input_image+' '+path_tmp+'data'+ext_data, self.verbose)
        if method_type == 'centerline':
            sct.run('isct_c3d '+self.method_val+' -o '+path_tmp+'/centerline.nii.gz')

        # go to tmp folder
        os.chdir(path_tmp)
        try:
            # convert to nii format
            sct.run('fslchfiletype NIFTI data', self.verbose)

            # Get dimensions of data
            nx, ny, nz, nt, px, py, pz, pt = sct.get_dimension('data.nii')
            sct.printv('  ' + str(nx) + ' x ' + str(ny) + ' x ' + str(nz)+ ' x ' + str(nt), self.verbose)

            # self.mask_method(nx, ny, nz, nt, px, py, pz, pt)

            # in case user input 4d data
            if nt != 1:
                sct.printv('WARNING in '+os.path.basename(__file__)+': Input image is 4d but output mask will 3D.', self.verbose, 'warning')
                # extract first volume to have 3d reference
                sct.run(fsloutput+'fslroi data data -0 1', self.verbose)

            # if method_type == 'coord':
            #     # parse to get coordinate
            #     coord = map(int, method_val.split('x'))
            #
            # if method_type == 'point':
            #     # get file name
            #     fname_point = method_val
            #     # extract coordinate of point
            #     sct.printv('\nExtract coordinate of point...', self.verbose)
            #     # TODO : Running outside script
            #     status, output = sct.run('sct_label_utils -i '+fname_point+' -t display-voxel', self.verbose)
            #     # parse to get coordinate
            #     coord = output[output.find('Position=')+10:-17].split(',')
            #
            # if method_type == 'center':
            #     # set coordinate at center of FOV
            #     coord = round(float(nx)/2), round(float(ny)/2)
            #
            # if method_type == 'centerline':
            #     # get name of centerline from user argument
            #     fname_centerline = 'centerline.nii.gz'
            # else:
            #     # generate volume with line along Z at coordinates 'coord'
            #     sct.printv('\nCreate line...', self.verbose)
            #     fname_centerline = self.create_line('data.nii', coord, nz)

            fname_centerline = self.mask_method(nx, ny, nz, nt, px, py, pz, pt)

            # create mask
            sct.printv('\nCreate mask...', self.verbose)
            centerline = nibabel.load(fname_centerline)  # open centerline
            hdr = centerline.get_header()  # get header
            hdr.set_data_dtype('uint8')  # set imagetype to uint8
            data_centerline = centerline.get_data()  # get centerline
            z_centerline = [iz for iz in range(0, nz, 1) if data_centerline[:, :, iz].any()]
            nz = len(z_centerline)
            # get center of mass of the centerline
            cx = [0] * nz
            cy = [0] * nz
            for iz in range(0, nz, 1):
                cx[iz], cy[iz] = ndimage.measurements.center_of_mass(numpy.array(data_centerline[:, :, z_centerline[iz]]))

            # create 2d masks
            file_mask = 'data_mask'
            for iz in range(nz):
                center = numpy.array([cx[iz], cy[iz]])
                mask2d = self.create_mask2d(center, self._shape, self._size, nx, ny)
                # Write NIFTI volumes
                img = nibabel.Nifti1Image(mask2d, None, hdr)
                nibabel.save(img, (file_mask+str(iz)+'.nii'))
            # merge along Z
            cmd = 'fslmerge -z mask '
            for iz in range(nz):
                cmd = cmd + file_mask+str(iz)+' '
            status, output = sct.run(cmd, self.verbose)

            # copy geometry
            sct.run(fsloutput+'fslcpgeom data mask', self.verbose)

        except:
            pass # TODO : Add the exception handler
        # come back to parent folder
        os.chdir('..')

        # Generate output files
        sct.printv('\nGenerate output files...', self.verbose)
        sct.generate_output_file(path_tmp+'mask.nii.gz', self._output_file)

        # Remove temporary files
        if self._rm_tmp_files == 1:
            sct.printv('\nRemove temporary files...', self.verbose)
            sct.run('rm -rf '+path_tmp, self.verbose)

        # to view results
        sct.printv('\nDone! To view results, type:', self.verbose)
        sct.printv('fslview '+self.input_image+' '+self._output_file+' -l Red -t 0.5 &', self.verbose, 'info')
        print

        self._result = Image(self._output_file)

        if self.produce_output == 0:
            try:
                os.remove(self._output_file)
            except OSError:
                sct.printv("WARNING : Couldn't remove output file. Either it is opened elsewhere or "
                           "it doesn't exist.", 0, 'warning')
        else:
            # Complete message
            sct.printv('\nDone! To view results, type:', self.verbose)
            sct.printv("fslview "+self._output_file+" &\n", self.verbose, 'info')

        return self._result


    # create_line
    # ==========================================================================================
    def create_line(self, fname, coord, nz):

        # duplicate volume (assumes input file is nifti)
        sct.run('cp '+fname+' line.nii', self.verbose)
        # set all voxels to zero
        sct.run('isct_c3d line.nii -scale 0 -o line.nii', self.verbose)
        # loop across z and create a voxel at a given XY coordinate
        cmd = 'sct_label_utils -i line.nii -o line.nii -t add -x '
        # for iz in range(nz):
        #     sct.run('sct_label_utils -i line.nii -o line.nii -t add -x '+str(int(coord[0]))+','+str(int(coord[1]))+','+str(iz)+',1', self.verbose)
        for iz in range(nz):
            if iz == nz-1:
                cmd += str(int(coord[0]))+','+str(int(coord[1]))+','+str(iz)+',1'
            else:
                cmd += str(int(coord[0]))+','+str(int(coord[1]))+','+str(iz)+',1:'

        sct.run(cmd, self.verbose)
        return 'line.nii'

    # create_mask2d
    # ==========================================================================================
    def create_mask2d(self, center, shape, size, nx, ny):

        # initialize 2d grid
        xx, yy = numpy.mgrid[:nx, :ny]
        mask2d = numpy.zeros((nx, ny))
        xc = center[0]
        yc = center[1]
        radius = round(float(size+1)/2)  # add 1 because the radius includes the center.

        if shape == 'box':
            mask2d[xc-radius:xc+radius+1, yc-radius:yc+radius+1] = 1

        elif shape == 'cylinder':
            mask2d = ((xx-xc)**2 + (yy-yc)**2 <= radius**2)*1

        elif shape == 'gaussian':
            sigma = float(radius)
            mask2d = numpy.exp(-(((xx-xc)**2)/(2*(sigma**2)) + ((yy-yc)**2)/(2*(sigma**2))))

        # import matplotlib.pyplot as plt
        # plt.imshow(mask2d)
        # plt.show()

        return mask2d

    # create_centerline
    # ==========================================================================================
    def create_centerline(self, coord, nz):
        sct.printv('\nCreate line...', self.verbose)
        return self.create_line('data.nii', coord, nz)

class MaskCoordinates(Mask):
    def __init__(self, input_image, output_file=None, method_value=None, shape=Mask.DEFAULT_SHAPE, size=Mask.DEFAULT_SIZE, verbose=1, rm_tmp_files=1, rm_output_file=0, produce_output=1):
        super(MaskCoordinates, self).__init__(input_image, output_file, method_value, shape, size, verbose, rm_tmp_files, rm_output_file, produce_output)

    def mask_method(self, nx, ny, nz, nt, px, py, pz, pt):
        coord = map(int, self.method_value.split('x'))
        return self.create_centerline(coord, nz)


class MaskPoint(Mask):
    def __init__(self, input_image, output_file=None, method_value=None, shape=Mask.DEFAULT_SHAPE, size=Mask.DEFAULT_SIZE, verbose=1, rm_tmp_files=1, rm_output_file=0, produce_output=1):
        super(MaskPoint, self).__init__(input_image, output_file, method_value, shape, size, verbose, rm_tmp_files, rm_output_file, produce_output)

    def mask_method(self, nx, ny, nz, nt, px, py, pz, pt):
        # get file name
        fname_point = self.method_value
        # extract coordinate of point
        sct.printv('\nExtract coordinate of point...', self.verbose)
        # TODO : Running outside script
        status, output = sct.run('sct_label_utils -i '+fname_point+' -t display-voxel', self.verbose)
        # parse to get coordinate
        coord = output[output.find('Position=')+10:-17].split(',')
        return self.create_centerline(coord, nz)


class MaskCenter(Mask):
    def __init__(self, input_image, output_file=None, method_value=None, shape=Mask.DEFAULT_SHAPE, size=Mask.DEFAULT_SIZE, verbose=1, rm_tmp_files=1, rm_output_file=0, produce_output=1):
        super(MaskCenter, self).__init__(input_image, output_file, method_value, shape, size, verbose, rm_tmp_files, rm_output_file, produce_output)

    def mask_method(self, nx, ny, nz, nt, px, py, pz, pt):
        coord = round(float(nx)/2), round(float(ny)/2)
        return self.create_centerline(coord, nz)


class MaskCenterline(Mask):
    def __init__(self, input_image, output_file=None, method_value=None, shape=Mask.DEFAULT_SHAPE, size=Mask.DEFAULT_SIZE, verbose=1, rm_tmp_files=1, rm_output_file=0, produce_output=1):
        super(MaskCenterline, self).__init__(input_image, output_file, method_value, shape, size, verbose, rm_tmp_files, rm_output_file, produce_output)

    def mask_method(self, nx, ny, nz, nt, px, py, pz, pt):
        return "centerline.nii.gz"
